### PR TITLE
Add support for Firefox (work in progress)

### PR DESCRIPTION
**Pull request by nc7s**
It runs on Firefox for now (dev edition 96.0b7), but `manifest.json` has been modified for Firefox only. Having separate `manifest.json`s for different browsers is necessary, and supposedly, also a `Makefile` or something similiar to package and upload to their respective extension stores.

### CODE REVIEW

**Comment by reorx** [comment]
统一 extension API 更好的写法是创造一个 class，将所有用到的 extension methods 都 rewrite 一遍:

```
class ExtensionAPI {
  constructor() {
    this.browser = browser
    if (isChrome) {
      this.browser = chrome
  }

  // Example for APIs with same references and arguments
  // chrome: chrome.storage.sync.get
  // firefox: browser.storage.sync.get
  storageSyncGet() {
    return this.browser.storage.sync.get.apply(arguments)
  }

  // Example for APIs with different references and arguments
  // chrome: chrome.scripting.executeScript
  // firefox: browser.tabs.executeScript
  executeScript() {
    if (isChrome) {
      return chrome.scripting.executeScript.apply(arguments)
    } else {
      return browser.tabs.executeScript.apply(arguments)
    }
  }
}

extapi = new ExtensionAPI()
```

File: popup.js [hunk 1]
```diff
@@ -4,16 +4,47 @@ const S_KEY = {
   defaultFormat: "defaultFormat",
 }
 
-
 /* Class definition */
 
+class ExtensionAPI {
```
Comment (reorx) [inline popup.js:9]: 注意这里以及其他地方的缩进，应统一使用 2 spaces

File: popup.js [hunk 2]
```diff
@@ -4,16 +4,47 @@ const S_KEY = {
...
+	constructor() {
+		if(typeof browser !== 'undefined') {
+			this.name = 'webext'
```
Comment (reorx) [inline popup.js:12]: API_NAME = {
  webext: 0,
  chrome: 1,
}

this.name = API_NAME.webext

File: popup.js [hunk 3]
```diff
@@ -4,16 +4,47 @@ const S_KEY = {
...
+}
+
 class App {
   constructor() {
     this.data = {
       title: null,
       url: null,
     }
     this.settings = {}
+    this.ext = new ExtensionAPI
```
Comment (reorx) [inline popup.js:47]: 声明全局变量 ext，不应该局限在 App 中

File: popup.js [hunk 4]
```diff
@@ -113,38 +144,31 @@ class App {
...
+          self.data.title = text;
+        }
+      }
       self.render();
-    }
+    };
+
+	try {
+		this.executeScript({ code: 'window.getSelection().toString()' }).then(afterSelect);
+	} catch {
```
Comment (reorx) [inline popup.js:169]: 非必要的异常处理

File: popup.js (hunk 2 above)
Comment (nc7s) [reply popup.js:12]: 这是什么对古早 enum 的执着

Comment (reorx) [reply popup.js:12]: 一点也不古早，enum 概念对任何语言都适用，越早写好处越多

**Pull request closed without merging**